### DISCUSSION TITLE
tutorials/shaders/your_first_shader/your_first_3d_shader

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#120] [User-contributed notes]
# tutorials/shaders/your_first_shader/your_first_3d_shader

You have decided to start writing your own custom Spatial shader. Maybe you saw a cool trick online that was done with shaders, or you have found that the StandardMaterial3D isn't quite meeting you...

https://docs.godotengine.org/en/stable/tutorials/shaders/your_first_shader/your_first_3d_shader.html

**Comment by mrlem** [comment]
I'm not maths /shader expert but shouldn't the normal calculation also be based on height_scale? It would seem logical, as this variable affects the triangles orientation. Unless that is handled under the hood by NORMAL_MAP?

**Reply by Sleepybean2** [reply]
This doesn't actually calculate your normals from the geometry at all. It's using the uv position in the texture and the color value to calculate a normal. That normal is accurate but not calculated directly from the geometry. So multiplying the tex_position will scale the uv position(and color value, but it's normalized) of the texture, meaning your normal will actually be for another position on the geometry. If you scale by 0.5, then you multiply the uv position by 2, resulting in a normal that is meant for a vertex half as far away.

**Reply by Sleepybean2** [reply]
sorry, by 1/2, not 2

**Reply by CoderJo-Pro** [reply]
Yes, normals are calculated according to the triangles, but you can also provide a normal map to tell it, "hey listen to me, this is the normal you're gonna use". So, the *As Normal* is exactly the handy utility that calculates the normal by noise value as height, and puts the normal inside the texture instead. This is a high-performance approach for enhancing quality without adding more expensive triangles.

**Comment by mitchellkager** [comment]
512x512 textures yielded a much grainier looking material as compared to the examples. I had more more similar results with lower resolutions e.g. 128x128 texture map and 64x64 normal map, with latter never exceeding the former.

**Comment by ParisMars** [comment]
If you want to smooth out the waves/reduce the bump density, you can divide (VERTEX.xz / 2.0) by another value (>1). You'll be sampling from points closer together in the noise texture, resulting in smoother height transitions. Note, this is not the same as reducing the height_scale value.

**Comment by Mika4334** [comment]
// ' *= '  - for proper light and shadow
NORMAL_MAP *= texture(normalmap,tex_position).xyz;

**Comment by PinkIr1s** [comment]
I found this way to generate noise in GDScript using FastNoiseLite, and then to send it to the shader on scene load, instead of doing it in the editor.
```gdscript
func _ready() -> void:
	var img = NoiseTexture2D.new()
	img.noise = FastNoiseLite.new()
	img.noise.noise_type = FastNoiseLite.TYPE_PERLIN
	img.noise.frequency = 0.005
	my_material.set_shader_parameter("my_noise", img)
```
